### PR TITLE
Support other implementations of Promises inside nuxtServerInit (currently only HTML5 Promises works correctly)

### PR DESCRIPTION
**Pull request by taldy**
Fix for using for exapmle Axios promises in nuxtServerInit() without wrapping it to HTML5 Promise

### CODE REVIEW

**Comment by Atinux** [comment]
Hi @taldy

Can you show an example where our code does not work?

**Comment by taldy** [comment]
@Atinux

```
// store/index.js

export const actions = {
  nuxtServerInit ({ commit }) {
    return axios.get('/api/products')
      .then(res => commit(SET_PRODUCTS, res.data))
  }
}
```

**Comment by taldy** [comment]
@Atinux I checked one more time and found that `async/await` solves my problem without any changes in Nuxt. 

But Vuex documentation describe returning promise from action as a legal way to use the system. That's why I thought this is a some kind of bug.

**Comment by any35** [comment]
I have same issue, maybe this code may be a better way to make sure the result is a promise. Not a object which has a property named 'then'
```
<% if (store) { %>
        let promise = (store._actions && store._actions.nuxtServerInit ? store.dispatch('nuxtServerInit', omit(getContext(context), 'redirect', 'error')) : null)
      promise = Promise.resolve(promise || undefined);
```

File: lib/app/server.js [hunk 1]
```diff
@@ -81,7 +81,7 @@ export default context => {
     // nuxtServerInit
     <% if (store) { %>
       let promise = (store._actions && store._actions.nuxtServerInit ? store.dispatch('nuxtServerInit', omit(getContext(context), 'redirect', 'error')) : null)
-      if (!(promise instanceof Promise)) promise = Promise.resolve()
+      if (!promise || (!(promise instanceof Promise) && (promise && typeof promise.then !== 'function'))) promise = Promise.resolve()
```
Comment (Atinux) [inline lib/app/server.js]: I believe the if can be simplify to this:
```js
if (!promise || (!(promise instanceof Promise) && typeof promise.then !== 'function'))
```

If `promise` does not pass the first check (`!promise`) it won't check the 2nd check.

Can you update this please?

**Review by atinux — CHANGES_REQUESTED** [review]

**Comment by Atinux** [comment]
What about updating this line as well @taldy to make every method which should return a Promise to not be only `Promise` instances? https://github.com/nuxt/nuxt.js/blob/master/lib/app/utils.js#L87

File: lib/app/server.js (hunk 1 above)
Comment (any35) [inline lib/app/server.js]: @Atinux 
axios or some other frameworks return  promise-like object, NOT a ES's Promise object. 
So the statement `promise instanceof Promise` will return `false`.
But a promise-like object behave like a promise and do works with ES's Promise object. 
Simply, just use Promise.resolve(xxx) will compatible with both promise and promise-like object
> ` promise = Promise.resolve(promise);`

BTW, to check if a object is a promise-like or not, `typeof promise.then !== 'function'`  is not enough.

**Review by atinux — APPROVED** [review]

**Pull request merged**